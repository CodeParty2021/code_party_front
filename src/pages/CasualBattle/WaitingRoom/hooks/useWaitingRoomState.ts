import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { CodeType, useFetchCodes } from "hooks/CodeAPIHooks/useFetchCodes";
import { UserState, UserAction } from "services/RoomSync/RoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";
import { User } from "services/user/user";

export type IResponse = {
  user: User | null;
  roomInfo: {
    roomId: string;
    host: UserState;
    memberKeys: string[];
    members: { [id: string]: UserState };
    actionKeys: string[];
    actions: { [id: string]: UserAction };
  };
  isHost: boolean;
  status?: "waiting" | "watching";
  ready: boolean;

  readyBtnHandler: () => void;
  readyBtnDisabled: boolean;
  exitBtnHandler: () => void;
  startBtnDisabled: boolean;
  startBtnHandler: () => void;
  kickUserHandler: (userId: string) => void;
  code: {
    codes: CodeType[];
    loading: boolean;
  };
  selectedCodeId: string;
  onChangeSelectedCodeId: (codeId: string) => void;
};

export const useWaitingRoomState = (): IResponse => {
  const {
    room,
    isHost,
    updateRoomInfo,
    updateMember,
    updateOtherMember,
    exitRoom,
  } = useRoomSync();
  const { data: codes = [], loading } = useFetchCodes();
  const [preRoomStatus, setPreRoomStatus] = useState<"waiting" | "watching">(
    "watching"
  );
  const [selectedCodeId, setSelectedCodeId] = useState("");
  const [ready, setReady] = useState(false);
  const [readyBtnDisabled, setReadyBtnDisabled] = useState(true);
  const [startBtnDisabled, setStartBtnDisabled] = useState(true);
  let { user } = useSelector((state: RootState) => state.user);

  const navigate = useNavigate();
  const dummyUser: UserState = {
    displayName: "",
    status: "disconnect",
    ready: false,
  };

  // ユーザ状態を更新
  useEffect(() => {
    updateMember({
      status: "waiting",
    });
  }, []);

  // DB上で退出処理が完了したら退出
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/casual-battle");
    }
  }, [room.isEntered]);

  // 準備ボタンの有効化
  useEffect(() => {
    //コードが選択されたら準備完了ボタンを押せるようにする
    if (selectedCodeId != "") {
      setReadyBtnDisabled(false);
    }
  }, [selectedCodeId]);

  // ルームステータスがwaitingからwatchingになったら観戦画面に遷移
  useEffect(() => {
    if (room.info) {
      if (preRoomStatus == "waiting" && room.info.status == "watching") {
        navigate("/casual-battle/result");
      }
      setPreRoomStatus(room.info?.status);
    }
  }, [room.info?.status]);

  // ready状態をDBに反映
  useEffect(() => {
    updateMember({
      ready: ready,
      codeId: selectedCodeId,
    });
  }, [ready]);

  // ホストの処理
  useEffect(() => {
    if (isHost) {
      if (room.info && room.info.status == "waiting") {
        // 全員が準備完了になったら開始ボタンを有効にする
        let newDisabled = false;
        for (let key of Object.keys(room.members)) {
          if (!room.members[key].ready) {
            newDisabled = true;
            break;
          }
        }
        setStartBtnDisabled(newDisabled);
      }
      // 全員がWaitingRoomに返ってきたらroomの状態をwaitingに更新する
      let newRoomStatus: "waiting" | "watching" = "waiting";
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "watching") {
          newRoomStatus = "watching";
        }
      }
      if (newRoomStatus == "waiting" && room.info?.status == "watching") {
        updateRoomInfo({
          status: "waiting",
          analyzingResult: null,
        });
      }
    }
  }, [room.members, isHost]);

  // kick処理 member.status = kickingの場合そのユーザをキックする
  useEffect(() => {
    if (user) {
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "kicking" && user.id == key) {
          exitRoom();
        }
      }
    }
  }, [room.members]);

  // ---- イベントハンドラ ---- //

  const _readyBtnHandler = () => {
    setReady(!ready);
  };

  const _exitBtnHandler = () => {
    exitRoom();
  };

  const _startBtnHandler = () => {
    updateRoomInfo({
      status: "watching",
    });
  };

  const _onChangeSelectedCodeId = (codeId: string) => {
    setSelectedCodeId(codeId);
  };

  const _kickUserBtnHandler = (userId: string) => {
    updateOtherMember(userId, { status: "kicking" });
  };

  return {
    user: user,
    roomInfo: {
      roomId: room.id ? room.id : "",
      host:
        room.info && room.info.host in room.members
          ? room.members[room.info.host]
          : dummyUser,
      memberKeys: room.sortedKeysOfMembers,
      members: room.members,
      actionKeys: room.sortedKeysOfActions,
      actions: room.actions,
    },
    isHost: isHost,
    status: room.info?.status,
    ready: ready,

    readyBtnHandler: _readyBtnHandler,
    readyBtnDisabled: readyBtnDisabled,
    exitBtnHandler: _exitBtnHandler,
    startBtnDisabled: startBtnDisabled,
    startBtnHandler: _startBtnHandler,
    kickUserHandler: _kickUserBtnHandler,
    code: {
      codes: codes,
      loading: loading,
    },
    selectedCodeId: selectedCodeId,
    onChangeSelectedCodeId: _onChangeSelectedCodeId,
  };
};
