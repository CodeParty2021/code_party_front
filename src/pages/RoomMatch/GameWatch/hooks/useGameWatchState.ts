import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";
import { useUnityGame } from "hooks/UnityGameHooks/useUnityGame";
import { UnityContext } from "react-unity-webgl";
import { useDummyLoading } from "hooks/DummyLoadingHooks/useDummyLoading";

export type IResponse = {
  unityContext: UnityContext;
  state: "Analyzing" | "Running" | "Finish" | "AnalyzingError";
  messageType: "Analyzing" | "AnalyzingError";
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: json, fetchJson } = useFetchJson();
  const { unityContext, startGame, unityStatus } = useUnityGame("SquarePaint");
  const [messageType, setMessageType] =
    useState<IResponse["messageType"]>("Analyzing");
  const { dummyLoadingState, startDummyLoad } = useDummyLoading(1000);
  const navigate = useNavigate();

  const state = room.info?.analyzingResult?.error
    ? "AnalyzingError"
    : !room.info?.analyzingResult?.resultId ||
      unityStatus.isLoading ||
      dummyLoadingState.isLoading
    ? "Analyzing"
    : unityStatus.isRunning
    ? "Running"
    : "Finish";

  // メッセージの操作
  useEffect(() => {
    if (state === "Analyzing") setMessageType("Analyzing");
    else if (state === "AnalyzingError") setMessageType("AnalyzingError");
  }, [state]);

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });

    // ダミーロード開始
    startDummyLoad();
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/room-match");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  // 観戦開始
  useEffect(() => {
    if (!unityStatus.isLoading && json) {
      startGame(json);
    }
  }, [unityStatus.isLoading, json]);

  const _exitBtnHandler = () => {
    navigate("/room-match/waiting-room");
  };

  return {
    state,
    messageType,
    unityContext,
    exitBtnHandler: _exitBtnHandler,
  };
};
