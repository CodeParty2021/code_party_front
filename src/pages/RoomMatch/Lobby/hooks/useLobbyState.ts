import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { RefObject, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  isLoading: boolean;
  errorMessage: string;
  roomIdRef: RefObject<HTMLInputElement>;
  createRoomHandler: () => void;
  createRoomDisabled: boolean;
  enterRoomHandler: () => Promise<void>;
  enterRoomDisabled: boolean;
  backButtonHandler: () => void;
};

export const useLobbyState = (): IResponse => {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>("");
  const [isProcessingAny, setIsProcessingAny] = useState<boolean>(false);
  const [succeedEnterRoom, setSucceedEnterRoom] = useState<boolean>(false);
  const { room, createRoom, enterRoom } = useRoomSync();
  const navigate = useNavigate();
  const roomIdRef = useRef<HTMLInputElement>(null!);

  const createEnterButtonDisabled = useMemo(() => (
    isProcessingAny || room.isEntered || succeedEnterRoom
  ), [isProcessingAny, room.isEntered]);

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  // ルーム
  const _createRoomHandler = useCallback(async () => {
    setIsLoading(true);
    setIsProcessingAny(true);
    await createRoom()
      .then(() => {
        setSucceedEnterRoom(true);
      })
      .catch(() => {
        navigate("/error");
      });
    setIsProcessingAny(false);
  }, []);

  const _enterRoomHandler = useCallback(async () => {
    const value = roomIdRef.current?.value;

    setIsProcessingAny(true);
    if (typeof value === "string") {
      if (value == "") {
        setErrorMessage("値を入力してください。");
      } else {
        await new Promise((resolve) => setTimeout(resolve, 500));
        await enterRoom(value).then(() => {
          setSucceedEnterRoom(true);
        }).catch((e) => {
          if (e.message == "roomId is empty") {
            setErrorMessage("値を入力してください。");
          } else if (e.message == "roomId is not found") {
            setErrorMessage("ルームIDが無効です。");
          } else if (e.message == "room is full") {
            setErrorMessage("ルームが満員です。");
          }
        });
      }
    } else {
      setErrorMessage("入力が不正です。");
    }
    setIsProcessingAny(false);
  }, []);

  // 戻るボタン
  const _backButtonHandler = useCallback(() => {
    navigate("/select-mode");
  }, []);

  return {
    isLoading,
    errorMessage,
    roomIdRef,
    createRoomHandler: _createRoomHandler,
    createRoomDisabled: createEnterButtonDisabled,
    enterRoomHandler: _enterRoomHandler,
    enterRoomDisabled: createEnterButtonDisabled,
    backButtonHandler: _backButtonHandler,
  };
};
