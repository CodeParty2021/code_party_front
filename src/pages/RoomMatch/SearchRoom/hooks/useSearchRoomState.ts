import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";

export type IResponse = {
  roomIdTextBoxValue: string;
  roomIdTextBoxChangeHandler: (value: string) => void;
  enterBtnDisabled: boolean;
  enterBtnClickHandler: () => void;
};

export const useSearchRoomState = (): IResponse => {
  const { room, enterRoom } = useRoomSync();
  const [roomId, setRoomId] = useState("");
  const [enterBtnDisabled, setEnterBtnDisabled] = useState(false);
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  const _roomIdTextBoxChangeHandler = (value: string) => {
    setRoomId(value);
  };

  const _enterBtnHandler = () => {
    setEnterBtnDisabled(true);
    enterRoom(roomId).catch(() => {
      setEnterBtnDisabled(false);
    });
  };

  return {
    roomIdTextBoxValue: roomId,
    roomIdTextBoxChangeHandler: _roomIdTextBoxChangeHandler,
    enterBtnDisabled: enterBtnDisabled,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
