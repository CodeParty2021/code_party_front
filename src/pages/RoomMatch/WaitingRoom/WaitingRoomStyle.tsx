import { Sword } from "components/icons";
import styled from "styled-components";
import { GRAY_100, GRAY_90 } from "styles/colors";
import { FlexGap } from "styles/FlexGap/FlexGap";

export const WaitingRoomStyle = styled.div`
  display: flex;
  padding: 128px;
  justify-content: space-around;
  align-items: center;
  height: 100vh;
`;

export const TitlePosition = styled.div`
  display: flex;
`;
export const Title = styled.h2`
  color: ${GRAY_100};
  font-size: 30px;
  font-weight: 700px;
  font-feature-settings: "palt" on;
  transform: matrix(1, 0, -0.08, 1, 0, 0);
  line-height: 30px;
  display: flex;
  align-items: center;
`;
export const SwordIcon = styled(Sword)`
  margin-right: 32px;
`;

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  ${FlexGap({ gap: "36px" })}
`;

export const PlayerBox = styled.div`
  display: flex;
  ${FlexGap({ gap: "24px" })}
`;
export const RightPanel = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`;
export const InfoPanel = styled.div`
  display: flex;
  ${FlexGap({ gap: "24px" })}
  align-items: flex-end;
`;
export const RoomInfo = styled.div`
  padding: 0 24px 0;
  font-family: "Noto Sans JP";
  font-style: normal;
  font-weight: 700;
  font-size: 24px;
  line-height: 35px;
  /* identical to box height */

  font-feature-settings: "palt" on;

  /* GRAY_90 */

  color: ${GRAY_90};
`;

export const RoomID = styled.div`
  font-family: "Noto Sans JP";
  font-style: normal;
  font-weight: 700;
  font-size: 48px;
  line-height: 48px;
  /* identical to box height */

  font-feature-settings: "palt" on;

  /* GRAY_90 */

  color: ${GRAY_90};

  transform: matrix(1, 0, -0.08, 1, 0, 0);
`;
export const Others = styled.div`
  display: flex;
  ${FlexGap({ gap: "24px" })}
`;
export const BottomPanel = styled.div`
  display: flex;
  justify-content: flex-end;
  ${FlexGap({ gap: "24px" })}
`;
