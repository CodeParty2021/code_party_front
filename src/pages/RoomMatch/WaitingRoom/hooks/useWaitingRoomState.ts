import { useEffect, useState, useMemo } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { CodeType, useFetchCodes } from "hooks/CodeAPIHooks/useFetchCodes";
import { UserState, UserAction } from "services/RoomSync/RoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";
import { User } from "services/user/user";

export type IResponse = {
  user: User | null;
  roomInfo: {
    roomId: string;
    invitationLink: string;
    host: UserState;
    memberKeys: string[];
    members: { [id: string]: UserState };
    actionKeys: string[];
    actions: { [id: string]: UserAction };
  };
  isHost: boolean;
  status?: "waiting" | "watching";
  ready: boolean;

  readyBtnHandler: () => void;
  readyBtnDisabled: boolean;
  exitBtnHandler: () => void;
  startBtnDisabled: boolean;
  startBtnHandler: () => void;
  isCopyBtnClicked: boolean;
  invitationBtnHandler: () => void;
  kickUserHandler: (userId: string) => void;
  code: {
    codes: CodeType[];
    loading: boolean;
  };
  selectedCode: CodeType | null;
  onChangeSelectedCode: (code: CodeType) => void;
  onStartToEditCode: () => void;
  showCodeSelectModal: boolean;
  openCodeSelectModal: () => void;
  closeCodeSelectModal: () => void;
};

export const useWaitingRoomState = (): IResponse => {
  const {
    room,
    isHost,
    updateRoomInfo,
    updateMember,
    updateOtherMember,
    exitRoom,
  } = useRoomSync();
  const { data: codes = [], loading } = useFetchCodes();
  const [preRoomStatus, setPreRoomStatus] = useState<"waiting" | "watching">(
    "watching"
  );
  const [selectedCode, setSelectedCode] = useState<CodeType | null>(null);
  const [ready, setReady] = useState(false);
  const [readyBtnDisabled, setReadyBtnDisabled] = useState(true);
  const [startBtnDisabled, setStartBtnDisabled] = useState(true);
  const [isCopyBtnClicked, setIsCopyBtnClicked] = useState(false);
  let { user } = useSelector((state: RootState) => state.user);
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const dummyUser: UserState = {
    displayName: "",
    status: "disconnect",
    ready: false,
  };

  const modalOn = useMemo(
    () => (searchParams.get("modal") === "on" ? true : false),
    []
  );

  // ユーザ状態を更新
  useEffect(() => {
    updateMember({
      status: "waiting",
    });
  }, []);

  // DB上で退出処理が完了したら退出
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/room-match");
    }
  }, [room.isEntered]);

  // 準備ボタンの有効化
  useEffect(() => {
    //コードが選択されたら準備完了ボタンを押せるようにする
    if (selectedCode != null) {
      setReadyBtnDisabled(false);
    }
  }, [selectedCode]);

  // ルームステータスがwaitingからwatchingになったら観戦画面に遷移
  useEffect(() => {
    if (room.info) {
      if (preRoomStatus == "waiting" && room.info.status == "watching") {
        navigate("/room-match/result");
      }
      setPreRoomStatus(room.info?.status);
    }
  }, [room.info?.status]);

  // ready状態をDBに反映
  useEffect(() => {
    updateMember({
      ready: ready,
      codeId: selectedCode ? selectedCode.id : "",
    });
  }, [ready]);

  // クエリパラメータにmodal=onを渡されたらモーダルを表示
  useEffect(() => {
    if (modalOn) {
      setShowCodeSelectModal(true);
    }
  }, [modalOn]);

  // ホストの処理
  useEffect(() => {
    if (isHost) {
      if (room.info && room.info.status == "waiting") {
        // 全員が準備完了になったら開始ボタンを有効にする
        let newDisabled = false;
        for (let key of Object.keys(room.members)) {
          if (!room.members[key].ready) {
            newDisabled = true;
            break;
          }
        }
        setStartBtnDisabled(newDisabled);
      }
      // 全員がWaitingRoomに返ってきたらroomの状態をwaitingに更新する
      let newRoomStatus: "waiting" | "watching" = "waiting";
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "watching") {
          newRoomStatus = "watching";
        }
      }
      if (newRoomStatus == "waiting" && room.info?.status == "watching") {
        updateRoomInfo({
          status: "waiting",
          analyzingResult: null,
        });
      }
    }
  }, [room.members, isHost]);

  // kick処理 member.status = kickingの場合そのユーザをキックする
  useEffect(() => {
    if (user) {
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "kicking" && user.id == key) {
          exitRoom();
        }
      }
    }
  }, [room.members]);

  // ---- イベントハンドラ ---- //

  const _readyBtnHandler = () => {
    setReady(!ready);
  };

  const _exitBtnHandler = () => {
    exitRoom();
  };

  const _startBtnHandler = () => {
    updateRoomInfo({
      status: "watching",
    });
  };

  const _onChangeSelectedCode = (code: CodeType) => {
    setSelectedCode(code);
  };

  const _onStartToEditCode = () => {
    if (selectedCode) {
      navigate(`/free-coding/${selectedCode.id}/selectCode`);
    }
  };

  //invitationURLのコピーボタン
  const _invitationBtnHandler = () => {
    if (room.invitationLink) {
      navigator.clipboard.writeText(room.invitationLink).then(() => {
        setIsCopyBtnClicked(true);
      });
    }
  };
  const _kickUserBtnHandler = (userId: string) => {
    updateOtherMember(userId, { status: "kicking" });
  };

  //モーダル表示非表示
  const [showCodeSelectModal, setShowCodeSelectModal] =
    useState<boolean>(false);
  const openCodeSelectModal = () => {
    setShowCodeSelectModal(true);
  };
  const closeCodeSelectModal = () => {
    setShowCodeSelectModal(false);
  };

  return {
    user: user,
    roomInfo: {
      roomId: room.id ? room.id : "",
      invitationLink: room.invitationLink ? room.invitationLink : "",
      host:
        room.info && room.info.host in room.members
          ? room.members[room.info.host]
          : dummyUser,
      memberKeys: room.sortedKeysOfMembers,
      members: room.members,
      actionKeys: room.sortedKeysOfActions,
      actions: room.actions,
    },
    isHost: isHost,
    status: room.info?.status,
    ready: ready,
    readyBtnHandler: _readyBtnHandler,
    readyBtnDisabled: readyBtnDisabled,
    exitBtnHandler: _exitBtnHandler,
    startBtnDisabled: startBtnDisabled,
    startBtnHandler: _startBtnHandler,
    isCopyBtnClicked: isCopyBtnClicked,
    invitationBtnHandler: _invitationBtnHandler,
    kickUserHandler: _kickUserBtnHandler,
    code: {
      codes: codes,
      loading: loading,
    },
    selectedCode: selectedCode,
    onChangeSelectedCode: _onChangeSelectedCode,
    onStartToEditCode: _onStartToEditCode,
    showCodeSelectModal,
    openCodeSelectModal,
    closeCodeSelectModal,
  };
};
