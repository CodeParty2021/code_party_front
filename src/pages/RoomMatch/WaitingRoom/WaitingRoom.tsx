import { Button } from "components/Button/Button";
import { Header } from "components/Header/Header";
import Link from "components/icons/Link";
import { PlayerPlateMe } from "components/PlayerPlate/Me/PlayerPlateMe";
import { PlayerPlateOther } from "components/PlayerPlate/Other/PlayerPlateOther";
import { StarBackground } from "components/StarBackground/StarBackground";
import React from "react";
import { SwordIcon } from "../Lobby/LobbyStyle";
import { CodeSelect } from "./components/CodeSelect";
import { useWaitingRoomState } from "./hooks/useWaitingRoomState";
import {
  BottomPanel,
  Container,
  InfoPanel,
  Others,
  PlayerBox,
  RightPanel,
  RoomID,
  RoomInfo,
  Title,
  TitlePosition,
  WaitingRoomStyle,
} from "./WaitingRoomStyle";

type Props = {};

export const RoomMatchWaitingRoom: React.FC<Props> = () => {
  const {
    user,
    roomInfo,
    isHost,
    ready,
    readyBtnHandler,
    exitBtnHandler,
    startBtnHandler,
    invitationBtnHandler,
    code,
    selectedCode,
    onChangeSelectedCode,
    openCodeSelectModal,
    closeCodeSelectModal,
    onStartToEditCode,
    showCodeSelectModal,
    startBtnDisabled,
    readyBtnDisabled,
  } = useWaitingRoomState();

  const labelize = (codeContent: string | undefined) => {
    if (codeContent == undefined) return "";
    const firstLine = codeContent.split("\n")[0];
    return firstLine;
  };
  const color: (
    memberKeys: string[],
    userId: string
  ) => "turquoise" | "leaf" | "orange" | "magenta" = (
    memberkeys: string[],
    userId: string
  ) => {
    switch (memberkeys.indexOf(userId)) {
      case 0:
        return "magenta";
      case 1:
        return "leaf";
      case 2:
        return "turquoise";
      case 3:
        return "orange";
    }
    return "magenta";
  };

  if (user) {
    return (
      <WaitingRoomStyle>
        {showCodeSelectModal && code && (
          <CodeSelect
            codeList={code.codes}
            onSelect={onChangeSelectedCode}
            onClose={closeCodeSelectModal}
            onStartToEdit={onStartToEditCode}
            selected={selectedCode}
            onReady={() => {
              if (!ready) readyBtnHandler();
            }}
            startToEditBtnDisabled={readyBtnDisabled}
            readyBtnDisabled={readyBtnDisabled}
          />
        )}

        <StarBackground />
        <Header backMessage="ロビーへ戻る" backButtonHandler={exitBtnHandler} />
        <Container>
          <TitlePosition>
            <SwordIcon></SwordIcon>
            <Title>フレンド対戦モード</Title>
          </TitlePosition>
          <PlayerBox>
            <PlayerPlateMe
              badge="称号無し"
              color={color(roomInfo.memberKeys, user.id)}
              onClickSelectCode={openCodeSelectModal}
              onClickSelectOtherCode={openCodeSelectModal}
              selectedCodeName={labelize(selectedCode?.codeContent)}
              status={ready ? "ready" : "default"}
              userName={user.displayName}
              userPhoto={user.picture}
            />

            <RightPanel>
              <InfoPanel>
                <RoomInfo>
                  ルームID
                  <RoomID>{roomInfo.roomId}</RoomID>
                </RoomInfo>

                <Button
                  Icon={Link}
                  color="black"
                  icon="left"
                  onClick={invitationBtnHandler}
                  size="S"
                  status="default"
                  value="招待リンクをコピー"
                />
              </InfoPanel>

              <Others>
                {roomInfo.memberKeys.map((key) => {
                  if (key == user.id) return;
                  return (
                    <div key={key}>
                      <PlayerPlateOther
                        badge="称号無し"
                        color={color(roomInfo.memberKeys, key)}
                        onClickChangeCPU={() => {}}
                        status={
                          roomInfo.members[key].ready
                            ? "ready"
                            : roomInfo.members[key].status === "disconnect"
                            ? "disconnecting"
                            : "default"
                        }
                        userName={roomInfo.members[key].displayName}
                        userPhoto={roomInfo.members[key].picture}
                        userType="user"
                      />
                    </div>
                  );
                })}
                {/* 待機中 */}
                {[1, 2, 3].map(
                  (i) =>
                    i >= roomInfo.memberKeys.length && (
                      <PlayerPlateOther
                        key={i}
                        badge=""
                        color="turquoise"
                        onClickChangeCPU={undefined}
                        status="waiting"
                        userName=""
                        userPhoto=""
                        userType="user"
                      />
                    )
                )}
              </Others>
            </RightPanel>
          </PlayerBox>
          <BottomPanel>
            <Button
              color="black"
              icon={null}
              onClick={exitBtnHandler}
              size="M"
              status="default"
              value="退出する"
            />
            {isHost && (
              <Button
                color="pink"
                icon={null}
                onClick={startBtnHandler}
                size="M"
                status={startBtnDisabled ? "disabled" : "default"}
                value="ゲームスタート"
              />
            )}
          </BottomPanel>
        </Container>
      </WaitingRoomStyle>
    );
  } else {
    return <div>ログインが必要です</div>;
  }
};
