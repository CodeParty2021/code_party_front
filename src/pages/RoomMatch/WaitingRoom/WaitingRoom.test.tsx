import React from "react";
import { shallow } from "enzyme";

import { RoomMatchWaitingRoom } from "./WaitingRoom";
import { IResponse, useWaitingRoomState } from "./hooks/useWaitingRoomState";

jest.mock("./hooks/useWaitingRoomState");

const useSelectorMock = useWaitingRoomState as jest.Mock;

const state: IResponse = {
  user: {
    id: "user-id-2",
    displayName: "user 2",
    email: "email",
    picture: "picture",
    jwt: "fewaf",
    isAnonymous: false,
  },
  roomInfo: {
    roomId: "room id",
    invitationLink:
      "http://localhost:3000/#/room-match/invitation/-MzESfkdu9Xmu8pkOAw5",
    host: {
      displayName: "host user",
      ready: false,
      status: "waiting",
      codeId: "hostCodeId",
    },
    memberKeys: ["host-user-id", "user-id-1", "user-id-2"],
    members: {
      "host-user-id": {
        displayName: "host user",
        ready: false,
        status: "waiting",
        codeId: "hostCodeId1",
      },
      "user-id-1": {
        displayName: "user 1",
        ready: true,
        status: "watching",
        codeId: "codeId1",
      },
      "user-id-2": {
        displayName: "user 2",
        ready: true,
        status: "disconnect",
        codeId: "codeId2",
      },
    },
    actionKeys: ["action id 1", "action id 2", "action id 3"],
    actions: {
      "action id 1": {
        actionId: 1,
        userId: "host user id",
      },
      "action id 2": {
        actionId: 2,
        userId: "user id 1",
      },
      "action id 3": {
        actionId: 3,
        userId: "user-id-2",
      },
    },
  },
  isHost: true,
  status: "waiting",
  ready: true,
  readyBtnHandler: jest.fn(),
  readyBtnDisabled: false,
  exitBtnHandler: jest.fn(),
  startBtnDisabled: false,
  startBtnHandler: jest.fn(),
  isCopyBtnClicked: false,
  invitationBtnHandler: jest.fn(),
  kickUserHandler: jest.fn(),
  code: {
    codes: [
      {
        id: "hostCodeId1",
        codeContent: "print('hello world)'",
        createdAt: "2022-02-16T05:05:46.315585+09:00",
        updatedAt: "2022-02-16T06:33:00.058575+09:00",
        language: "1",
        step: "3",
        user: "host user id",
      },
      {
        id: "hostCodeId2",
        codeContent: "alert('hello world)'",
        createdAt: "2022-02-16T10:00:46.315585+09:00",
        updatedAt: "2022-02-16T11:10:00.058575+09:00",
        language: "1",
        step: "3",
        user: "host user id",
      },
    ],
    loading: false,
  },
  selectedCodeId: "codeId1",
  onChangeSelectedCodeId: jest.fn(),
};

describe("<RoomMatchWaitingRoom />", () => {
  beforeEach(() => {
    useSelectorMock.mockReturnValue({
      ...state,
    });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("snapshot test", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it("start button test", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    const btn = wrapper.find("#start-btn");
    btn.simulate("click");

    expect(state.startBtnHandler).toHaveBeenCalled();
  });

  it("kick user button test", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    const btn = wrapper.find("#kick-btn-user-id-1");
    btn.simulate("click");

    expect(state.kickUserHandler).toHaveBeenCalled();
  });

  it("exit button test", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    const btn = wrapper.find("#exit-btn");
    btn.simulate("click");

    expect(state.exitBtnHandler).toHaveBeenCalled();
  });

  it("ready button test", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    const btn = wrapper.find("#ready-btn");
    btn.simulate("click");

    expect(state.readyBtnHandler).toHaveBeenCalled();
  });

  it("inputが変化した時にonChangeSelectedCodeIdが呼ばれるか", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    wrapper.find("input[value='hostCodeId1']").simulate("change", {
      target: {
        value: "selectedCodeId",
      },
    });
    expect(state.onChangeSelectedCodeId).toBeCalledTimes(1);
    expect(state.onChangeSelectedCodeId).lastCalledWith("selectedCodeId");
  });

  it("copy button test", () => {
    const wrapper = shallow(<RoomMatchWaitingRoom />);

    const btn = wrapper.find("#invitation-btn");
    btn.simulate("click");

    expect(state.invitationBtnHandler).toHaveBeenCalled();
  });
});
