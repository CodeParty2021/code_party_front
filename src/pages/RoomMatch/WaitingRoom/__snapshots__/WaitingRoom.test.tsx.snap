// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomMatchWaitingRoom /> snapshot test 1`] = `
Array [
  <styled.div>
    <StarBackground />
    <Header
      backButtonHandler={[MockFunction]}
      backMessage="ロビーへ戻る"
    />
    <styled.div>
      <styled.div>
        <Styled(Sword) />
        <styled.h2>
          フレンド対戦モード
        </styled.h2>
      </styled.div>
      <styled.div>
        <PlayerPlateMe
          badge="称号無し"
          color="turquoise"
          onClickSelectCode={[Function]}
          onClickSelectOtherCode={[Function]}
          selectedCodeName="print('hello world)'"
          status="ready"
          userName="user 2"
          userPhoto="picture"
        />
        <styled.div>
          <styled.div>
            <styled.div>
              ルームID
              <styled.div>
                room id
              </styled.div>
            </styled.div>
            <Button
              Icon={[Function]}
              color="black"
              icon="left"
              onClick={[MockFunction]}
              size="S"
              status="default"
              value="招待リンクをコピー"
            />
          </styled.div>
          <styled.div>
            <div>
              <PlayerPlateOther
                badge="称号無し"
                color="magenta"
                onClickChangeCPU={[Function]}
                status="default"
                userName="host user"
                userType="user"
              />
            </div>
            <div>
              <PlayerPlateOther
                badge="称号無し"
                color="leaf"
                onClickChangeCPU={[Function]}
                status="ready"
                userName="user 1"
                userType="user"
              />
            </div>
            <PlayerPlateOther
              badge=""
              color="turquoise"
              status="waiting"
              userName=""
              userPhoto=""
              userType="user"
            />
          </styled.div>
        </styled.div>
      </styled.div>
      <styled.div>
        <Button
          color="black"
          icon={null}
          onClick={[MockFunction]}
          size="M"
          status="default"
          value="退出する"
        />
        <Button
          color="pink"
          icon={null}
          onClick={[MockFunction]}
          size="M"
          status="default"
          value="ゲームスタート"
        />
      </styled.div>
    </styled.div>
  </styled.div>,
]
`;
